fabtests/benchmarks/rdm_bw_mt.c

```diff
@@ -0,0 +1,734 @@
+ * WARNING: Not all options are supported in this test!
+ */
+
+#include <rdma/fi_cm.h>
+
+#include "shared.h"
+#include "benchmark_shared.h"
+#include "hmem.h"
+
+#define BUFFER_SIZE 1024
+static char buffer[BUFFER_SIZE];
+
+static int num_eps = 1;
+static bool bidir = false;
+pthread_barrier_t barrier;
+
+struct thread_args {
+	pthread_t thread;
+	fi_addr_t fiaddr;
+	struct fid_domain *domain;
+	struct fid_ep *ep;
+	struct fid_cq *cq;
+	struct fid_cntr *tx_cntr;
+	struct fid_cntr *rx_cntr;
+	struct fid_av *av;
+	struct fid_eq *eq;
+	struct fid_mr *tx_mr;
+	struct fid_mr *rx_mr;
+	void *tx_mr_desc;
+	void *rx_mr_desc;
+	struct fi_context2 send_ctx;
+	struct fi_context2 recv_ctx;
+	size_t size;
+	char *tx_buf;
+	char *rx_buf;
+	int id;
+	int ret;
+};
+
+static struct thread_args *targs = NULL;
+
+static void cleanup_ofi(void)
+{
+	int ret;
+	int i;
+
+	for (i = 0; i < num_eps; i++) {
+		if (targs[i].ep) {
+			ret = fi_close(&targs[i].ep->fid);
+			if (ret)
+				printf("fi_close(ep[%d]) failed: %d\n", i, ret);
+		}
+
+		if (targs[i].cq) {
+			ret = fi_close(&targs[i].cq->fid);
+			if (ret)
+				printf("fi_close(cq[%d]) failed: %d\n", i, ret);
+		}
+
+		if (targs[i].tx_cntr) {
+			ret = fi_close(&targs[i].tx_cntr->fid);
+			if (ret)
+				printf("fi_close(tx_cntr) failed: %d\n", ret);
+		}
+
+		if (targs[i].rx_cntr) {
+			ret = fi_close(&targs[i].rx_cntr->fid);
+			if (ret)
+				printf("fi_close(rx_cntr) failed: %d\n", ret);
+		}
+
+		if (targs[i].av) {
+			ret = fi_close(&targs[i].av->fid);
+			if (ret)
+				printf("fi_close(av[%d]) failed: %d\n", i, ret);
+		}
+
+		if (targs[i].domain) {
+			ret = fi_close(&targs[i].domain->fid);
+			if (ret)
+				printf("fi_close(domain[%d]) failed: %d\n", i,
+					ret);
+		}
+
+		if (targs[i].eq) {
+			ret = fi_close(&targs[i].eq->fid);
+			if (ret)
+				printf("fi_close(eq[%d]) failed: %d\n", i, ret);
+		}
+	}
+
+	if (fabric) {
+		fi_close(&fabric->fid);
+		if (ret)
+			printf("fi_close(fabric) failed: %d\n", ret);
+	}
+
+	if (fi)
+		fi_freeinfo(fi);
+	if (hints)
+		fi_freeinfo(hints);
+	if (targs)
+		free(targs);
+}
+
+static int init_av(int i)
+{
+	int ret;
+	size_t len = BUFFER_SIZE;
+	char print_buf[BUFFER_SIZE];
+
+	memset(print_buf, 0, BUFFER_SIZE);
+
+	ret = fi_getname(&targs[i].ep->fid, buffer, &len);
+	if (ret) {
+		printf("fi_getname failed: %d\n", ret);
+		return ret;
+	}
+
+	len = BUFFER_SIZE;
+
+	ret = ft_sock_send(oob_sock, buffer, len);
+	if (ret) {
+		printf("ft_sock_send failed: %d\n", ret);
+		return ret;
+	}
+
+	ret = ft_sock_recv(oob_sock, buffer, len);
+	if (ret) {
+		printf("ft_sock_recv failed: %d\n", ret);
+		return ret;
+	}
+
+	ret = fi_av_insert(targs[i].av, buffer, 1, &targs[i].fiaddr, 0, NULL);
+	if (ret != 1) {
+		printf("fi_av_insert failed: %d\n", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int init_ofi(void)
+{
+	struct fi_cq_attr cq_attr = {0};
+	struct fi_av_attr av_attr = {0};
+	struct fi_eq_attr eq_attr = {0};
+	struct fi_cntr_attr cntr_attr = {0};
```

Comment:
These are cleared by `memset` before use so there is no need to initialize here.